tests/testthat/test-S4.R

```diff
@@ -33,3 +36,75 @@ test_that("converts S4 representation of S3 classes to R7 representation", {
 test_that("errors on non-S4 classes", {
   expect_snapshot(S4_to_R7_class(1), error = TRUE)
 })
+
+
+describe("S4_class_dispatch", {
+  it("returns name of base class", {
+    on.exit(S4_remove_classes("Foo1"))
+    setClass("Foo1", slots = list("x" = "numeric"), where = globalenv())
+    expect_equal(S4_class_dispatch("Foo1"), "S4/Foo1")
+  })
+
+  it("respects single inheritance hierarchy", {
+    on.exit(S4_remove_classes(c("Foo1", "Foo2","Foo3")))
+
+    setClass("Foo1", slots = list("x" = "numeric"), where = globalenv())
+    setClass("Foo2", contains = "Foo1", where = globalenv())
+    setClass("Foo3", contains = "Foo2", where = globalenv())
+    expect_equal(S4_class_dispatch("Foo3"), c("S4/Foo3", "S4/Foo2", "S4/Foo1"))
+  })
+
+  it("performs breadth first search for multiple dispatch", {
+    on.exit(S4_remove_classes(c("Foo1a", "Foo1b","Foo2a", "Foo2b", "Foo3")))
+    setClass("Foo1a", slots = list("x" = "numeric"), where = globalenv())
+    setClass("Foo1b", contains = "Foo1a", where = globalenv())
+    setClass("Foo2a", slots = list("x" = "numeric"), where = globalenv())
+    setClass("Foo2b", contains = "Foo2a", where = globalenv())
+    setClass("Foo3", contains = c("Foo1b", "Foo2b"), where = globalenv())
+    expect_equal(
+      S4_class_dispatch("Foo3"),
+      c("S4/Foo3", "S4/Foo1b", "S4/Foo2b", "S4/Foo1a", "S4/Foo2a")
+    )
+  })
+
+  it("handles extensions of base classes", {
+    on.exit(S4_remove_classes("Foo1"))
+    setClass("Foo1", contains = "character", where = globalenv())
+    expect_equal(S4_class_dispatch("Foo1"), c("S4/Foo1", "character"))
+  })
+
+  it("handles extensions of S3 classes", {
+    on.exit(S4_remove_classes(c("Soo1", "Foo2", "Foo3")))
+
+    setOldClass(c("Soo1", "Soo"), where = globalenv())
+    setClass("Foo2", contains = "Soo1", where = globalenv())
+    setClass("Foo3", contains = "Foo2", where = globalenv())
+    expect_equal(S4_class_dispatch("Foo3"), c("S4/Foo3", "S4/Foo2", "Soo1", "Soo"))
+  })
+
+  it("ignores unions", {
+    on.exit(S4_remove_classes(c("Foo1", "Foo2", "Foo3")))
+
+    setClass("Foo1", slots = list("x" = "numeric"), where = globalenv())
+    setClass("Foo2", slots = list("x" = "numeric"), where = globalenv())
+    setClassUnion("Foo3", c("Foo1", "Foo2"), where = globalenv())
+
+    expect_equal(S4_class_dispatch("Foo1"), "S4/Foo1")
+    expect_equal(S4_class_dispatch("Foo2"), "S4/Foo2")
```

Comment:
What does `S4_class_dispatch("Foo3")` return? I'm assuming that is not allowed in some way, maybe the idea is that we shouldn't check that here because it is handled at registration time instead

Reply:
Yeah, R7 will never see an instance of that type so we don't need to test it.